pkg/store/store.go

```diff
@@ -116,6 +124,121 @@ type CacheStores struct {
 	KnativeIngress cache.Store
 }
 
+// NewCacheStores is a convenience function for CacheStores to initialize all attributes with new cache stores
+func NewCacheStores() (c CacheStores) {
+	c.ClusterPlugin = cache.NewStore(keyFunc)
+	c.Configuration = cache.NewStore(keyFunc)
+	c.Consumer = cache.NewStore(keyFunc)
+	c.Endpoint = cache.NewStore(keyFunc)
+	c.IngressV1 = cache.NewStore(keyFunc)
+	c.IngressV1beta1 = cache.NewStore(keyFunc)
+	c.KnativeIngress = cache.NewStore(keyFunc)
+	c.Plugin = cache.NewStore(keyFunc)
+	c.Secret = cache.NewStore(keyFunc)
+	c.Service = cache.NewStore(keyFunc)
+	c.TCPIngress = cache.NewStore(keyFunc)
+	c.UDPIngress = cache.NewStore(keyFunc)
+	return
+}
+
+// NewCacheStoresFromObjYAML provides a new CacheStores object given any number of byte arrays containing
+// YAML Kubernetes objects. An error is returned if any provided YAML was not a valid Kubernetes object.
+func NewCacheStoresFromObjYAML(objs ...[]byte) (c CacheStores, err error) {
+	kobjs := make([]runtime.Object, 0, len(objs))
+	for _, yaml := range objs {
+		sr := serializer.NewYAMLSerializer(
+			yamlserializer.DefaultMetaFactory,
+			unstructuredscheme.NewUnstructuredCreator(),
+			unstructuredscheme.NewUnstructuredObjectTyper(),
+		)
+		kobj, _, decodeErr := sr.Decode(yaml, nil, nil)
+		if err = decodeErr; err != nil {
+			return
+		}
+		kobjs = append(kobjs, kobj)
+	}
+	return NewCacheStoresFromObjs(kobjs...)
+}
+
+// NewCacheStoresFromObjs provides a new CacheStores object given any number of Kubernetes
+// objects that should be pre-populated. This function will sort objects into the appropriate
+// sub-storage (e.g. IngressV1, TCPIngress, e.t.c.) but will produce an error if any of the
+// input objects are erroneous or otherwise unusable as Kubernetes objects.
+func NewCacheStoresFromObjs(objs ...runtime.Object) (c CacheStores, err error) {
+	c = NewCacheStores()
+	for _, obj := range objs {
+		switch gvk := obj.GetObjectKind().GroupVersionKind(); gvk {
```

Comment:
I feel intimidated by this `switch` block :joy:.

How about splitting it into two smaller functions, each with a narrower responsibility:

```go
func makeObjectFromGVK(gvk GroupVersionKind) (runtime.Object, error) {
  switch gvk {
  case extensions.SchemeGroupVersion.WithKind("Ingress"):
    return &extensions.Ingress{}
  case networkingv1.SchemeGroupVersion.WithKind("Ingress"):
    return &networkingv1.Ingress{}
  // ... etc
  }
}
```

```go
func (c *CacheStores) Add(obj runtime.Object) error {
  switch obj := obj.(type) {
  case *extensions.Ingress:
    return c.IngressV1Beta1.Add(obj)
  case *networkingv1.Ingress:
    return c.IngressV1.Add(obj)
  // ...
  }
}
```

so that `NewCacheStoresFromObjs` becomes something like:

```go
func NewCacheStoresFromObjs(objs ...runtime.Object) (c CacheStores, err error) {
	c = NewCacheStores()
  for _, obj := range objs {
    var typedObj runtime.Object
    var err error
    typedObj, err = makeObjectFromGVK(obj.GetObjectKind().GroupVersionKind())
    // handle err

    if err = convUnstructuredObj(&obj, &typedObj); err != nil {
      // handle err
    }

    if err := c.Add(typedObj); err != nil {
      // handle err
    }
  }
```

Reply:
Sounds good, 2cf36d39b2e9578978655b0e75ba50c336eaedb9 makes these changes :+1: